server/msteams/client.go

```diff
@@ -243,27 +246,50 @@ func (at AccessToken) GetToken(_ context.Context, _ policy.TokenRequestOptions)
 	}, nil
 }
 
+type MetricsMiddleware struct {
+	metricsService metrics.Metrics
+}
+
+func (m *MetricsMiddleware) Intercept(pipeline khttp.Pipeline, middlewareIndex int, req *http.Request) (*http.Response, error) {
+	if m.metricsService != nil {
+		// Using this middleware results in the use of "users/me-token-to-replace" in the URL instead of "me"
+		// so we are fixing that.
+		// TODO: Explore why this happens and find a more appropriate solution for this
+		req.URL.Path = strings.Replace(req.URL.Path, "users/me-token-to-replace", "me", 1)
+
+		now := time.Now()
+		response, err := pipeline.Next(req, middlewareIndex)
+		elapsed := float64(time.Since(now)) / float64(time.Second)
+		m.metricsService.ObserveMSGraphAPIEndpointDuration(req.URL.RequestURI(), req.Method, fmt.Sprint(response.StatusCode), elapsed)
```

Comment:
What do you think we should log then? @jespino

Reply:
One not fully realible solution would be to parse the URL and replace anything that matches with an ID in the MSGraphAPI wit shomething like '<id>', that way all the urls are going to be identical, and the counting is going to work well.

Reply:
The thing is this is not a "log", this is a metric, and because it is a metric and how prometheus work, adding one metric entry for each call would flood our metrics system and generate problems. The whole point of a metric here is to have an aggregated number of calls by "endpoint" not by specific resource, because there can be millions of different resources, but the number of endpoints is limited. My whole point is we need to have something that clearly refers to specific endpoints, not to specific resources. What is the solution? maybe the ID replacement, maybe other thing like taking tracebacks and logging the specific method called from msteams.Client (I think that can have a bad performance)...

I don't have a clear opinion here I would have to investigate it by myself to have a better option defined there. But anything that tracks general endpoints instead of specific resources should do the trick.

Reply:
I'm thinking about using regex to remove the ids from URL. What do you think? @jespino @lieut-data

Reply:
I'm not a big fan of regular expressions, specially for something that you can do in a more sane way. an URI it should be a "path", so I would use "path pasing" instead, and then I would match each entry in the path with a regular expression. Not just a regular expresion over the whole URI, that I would say that is more error prone. The less we relay in a complex regular expression, the better.

Reply:
Unfortunately, I don't think regular expressions alone will give confidence that we don't take down our Prometheus instance. I'd rather we hold off merging until we have a way of explicitly mapping the URL to a fixed set of identifiers. (Is there a "reverse lookup" for OpenAPI specs?)